File: Java/src/LinkedLists/LRUCache.java
package LinkedLists;

import DS.DoublyLinkedListNode;
import java.util.HashMap;

/*
    // Definition of DoublyLinkedListNode:
    public class DoublyLinkedListNode {
        public int key;
        public int val;
        public DoublyLinkedListNode next;
        public DoublyLinkedListNode prev;
        public DoublyLinkedListNode(int key, int val) {
            this.key = key;
            this.val = val;
        }
    }
 */

public class LRUCache {
    public int capacity;
    public HashMap<Integer, DoublyLinkedListNode> cache;
    public DoublyLinkedListNode head;
    public DoublyLinkedListNode tail;
    public LRUCache(int capacity) {
        this.capacity = capacity;
        // A hash map that maps keys to nodes.
        cache = new HashMap<>();
        // Initialize the head and tail dummy nodes and connect them to
        // each other to establish a basic two-node doubly linked list.
        head = new DoublyLinkedListNode(-1, -1);
        tail = new DoublyLinkedListNode(-1, -1);
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }
        else {
            // To make this key the most recently used, remove its node and
            // re-add it to the tail of the linked list.
            remove(cache.get(key));
            addToTail(cache.get(key));
            return cache.get(key).val;
        }
    }

    public void put(int key, int val) {
        // If a node with this key already exists, remove it from the
        // linked list.
        if (cache.containsKey(key)) {
            remove(cache.get(key));
        }
        DoublyLinkedListNode node = new DoublyLinkedListNode(key, val);
        cache.put(key, node);
        // Remove the least recently used node from the cache if adding
        // this new node will result in an overflow.
        if (cache.size() > capacity) {
            cache.remove(head.next.key);
            remove(head.next);
        }
        addToTail(node);
    }

    private void addToTail(DoublyLinkedListNode node) {
        DoublyLinkedListNode prevNode = tail.prev;
        node.prev = prevNode;
        node.next = tail;
        prevNode.next = node;
        tail.prev = node;
    }

    private void remove(DoublyLinkedListNode node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

}

File: csharp/Linked Lists/LRUCache.cs
using ds.DoublyLinkedListNode;

class DoublyLinkedListNode
{
	public DoublyLinkedListNode(int key, int val)
	{
		Key = key;
		Val = val;
		Prev = null;
		Next = null;
	}

	public int Key { get; set; }

	public int Val { get; set; }

	public DoublyLinkedListNode Prev { get; set; }

	public DoublyLinkedListNode Next { get; set; }
}

public class LRUCache
{
	private readonly int _capacity;
	private readonly Dictionary<int, DoublyLinkedListNode> _hashMap;
	private readonly DoublyLinkedListNode _head;
	private readonly DoublyLinkedListNode _tail;

	public LRUCache(int capacity)
    {
		_capacity = capacity;

		// A hash map that maps keys to nodes
		_hashMap = new Dictionary<int, DoublyLinkedListNode>();
		
		// Initialize the head and tail dummy nodes and connect them to 
		// each other to establish a basic two-node doubly linked list.
		_head = new DoublyLinkedListNode(-1, -1);
		_tail = new DoublyLinkedListNode(-1, -1);		
		_head.Next = _tail;
		_tail.Prev = _head;
	}

	public int Get(int key)
	{
		if (!_hashMap.ContainsKey(key))
			return -1;

		// To make this key the most recently used, remove its node and 
		// re-add it to the tail of the linked list.
		removeNode(_hashMap[key]);
		addToTail(_hashMap[key]);

		return _hashMap[key].Val;
	}

	public void Put(int key, int value)
	{
		// If a node with this key already exists, remove it from the linked list.
		if (_hashMap.ContainsKey(key))
			removeNode(_hashMap[key]);

		DoublyLinkedListNode node = new DoublyLinkedListNode(key, value);
		_hashMap[key] = node;

		// Remove the least recently used node from the cache if adding
		// this new node will result in an overflow.
		if (_hashMap.Count > _capacity)
		{
			_hashMap.Remove(_head.Next.Key);
			removeNode(_head.Next);
		}

		addToTail(node);
	}

	private void addToTail(DoublyLinkedListNode node)
	{
		DoublyLinkedListNode prevNode = _tail.Prev;
		
		node.Prev = prevNode;
		node.Next = _tail;

		prevNode.Next = node;
		_tail.Prev = node;
	}

	private void removeNode(DoublyLinkedListNode node)
	{
		node.Prev.Next = node.Next;
		node.Next.Prev = node.Prev;
	}
}
